Part1_InterruptControllersEvolution.html

```diff
@@ -104,79 +104,78 @@ <h3>PIC</h3>
+	<p>The next step in architecture development was named <a href="https://www.intel.com/content/dam/doc/specification-update/64-architecture-x2apic-specification.pdf">x2APIC</a>. The number of possible CPUs in the system was increased to 2^32. These controllers can work in a backwards compatibility mode with xAPIC, or they can work in the new x2APIC mode. In this new mode controller programming is not done through MMIO, but through MSR registers (which are much faster). According to <a href="https://blahg.josefsipek.net/?tag=xAPIC">this link</a>, IOMMU support is necessary for this mode.<p/>
 
-	<p>It is worth to notice that it is possible to have many I/O APIC controllers in the system. For example, one for 24 interrupts in a southbridge and the other one for 32 interrupts in a northbridge. In the context of I/O APIC, interrupts usually titled as GSI (Global System Interrupt). So, the forementioned system has GSIs 0-55.<p/>
+	<p>It is worthwhile to note that it is possible to have several I/O APIC controllers in the system. For example, one for 24 interrupts in a southbridge and the other one for 32 interrupts in a northbridge. In the context of I/O APIC, interrupts are usually called GSI (Global System Interrupt). So, the forementioned system has GSIs 0-55.<p/>
 
-	<p>Does the CPU have an internal LAPIC and which APIC architecture exactly? It is possible to get answers to these questions by inspecting bit-flags from CPUID.
-	To help OS to discover LAPIC and I/O APIC, BIOS should present information about them either through a MPtable (old method) or through an ACPI table (a MADT table in this case). Besides common information, both the MPtable and the ACPI (in this case a DSDT table) should contain info about the interrupt routing. This means information about which device uses which interrupt line (a $PIR table analogy).<p/>
+	<p>How can we determine if a CPU has an internal LAPIC and which APIC architecture it supports? It is possible to answer these questions by inspecting bit-flags from CPUID.
+	To help the OS discover LAPIC and I/O APIC, the BIOS presents information about them either through an MPtable (old method) or through an ACPI table (a MADT table in this case). Besides common information, both the MPtable and the ACPI (in this case a DSDT table) should contain information about the interrupt routing. This means information about which device uses which interrupt line (similar to the $PIR table).<p/>

```

Comment:
"the BIOS presents information about them". Why the variant with "the BIOS should present information about them" is no good? Cause buggy BIOS could not have this information

Reply:
That's a good point, "should" is good then.

Reply:
```suggestion
	To help the OS discover LAPIC and I/O APIC, the BIOS should presents information about them either through an MPtable (old method) or through an ACPI table (a MADT table in this case). Besides common information, both the MPtable and the ACPI (in this case a DSDT table) should contain information about the interrupt routing. This means information about which device uses which interrupt line (similar to the $PIR table).<p/>
```

Reply:
It should be "present" in this case

Reply:
```suggestion
	To help the OS discover LAPIC and I/O APIC, the BIOS should present information about them either through an MPtable (old method) or through an ACPI table (a MADT table in this case). Besides common information, both the MPtable and the ACPI (in this case a DSDT table) should contain information about the interrupt routing. This means information about which device uses which interrupt line (similar to the $PIR table).<p/>
```